### PR TITLE
[Merged by Bors] - Add timer-based common run conditions (`on_timer` and `on_fixed_timer`)

### LINKED ISSUE #7864: Add timer-based common run conditions
## What problem does this solve or what need does it fill?

Users often want to run a system (like logging, polling or factory ticks) on a regular timer. Currently, this involves a fair bit of type-level boilerplate to define a resource type, or clever use of either run conditions or `Local` system variables.

## What solution would you like?

Create a run condition that stores a `Timer` in its `Local`, with a convenient method to generate such a run conditions.

Ideally this would look like:
```rust
app.add_system(my_system.run_if(on_timer(custom_duration))`
```

There should also be a `on_fixed_timer` variant, which uses elapsed `FixedTime`.

See the existing common run conditions in `bevy_ecs` and `bevy_input` for how to do so.

## What alternative(s) have you considered?

We could add multiple fixed time steps again (which were previously used for this), but these are not equivalent and this previous pattern was incorrect. See #7637.

## Additional context

Credit to @maniwani for the idea: I just made the ticket ;)

### PR DESCRIPTION
**Pull request by jabuwu**
# Objective

Fixes #7864

## Solution

Add the run conditions described in the issue. Also needed to add `bevy` as a dev dependency to `bevy_time` so the doctests can run.

---

## Changelog

- Add `on_timer` and `on_fixed_timer` run conditions

### CODE REVIEW

**Comment by mockersf** [comment]
I would prefer to not add the full bevy dependency just for doc tests...
`bevy_app` is already in the dependencies, that should be enough to write an example that compiles.

For an example, see 
https://github.com/bevyengine/bevy/blob/91ff78243916e622a7db38ae0f17237963a814e7/crates/bevy_log/src/lib.rs#L49-L61

Also you can hide lines not relevant by adding a `#` first

**Comment by jabuwu** [comment]
> I would prefer to not add the full bevy dependency just for doc tests...

Good call! I made the change. The only disadvantage is that the sample no longer includes the `use` statement, and `on_timer`/`on_fixed_timer` are not in the prelude (following `bevy_input`'s lead).

File: crates/bevy_time/src/common_conditions.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+/// delta time).
+///
+/// For more accurate timers, use the [`Timer`] class directly (see
+/// [`Timer::times_finished_this_tick`] to address the problem mentioned above), or
+/// use fixed timesteps that allow systems to run multiple times per frame.
+pub fn on_timer(duration: Duration) -> impl FnMut(Res<Time>) -> bool {
+    let mut timer = Timer::new(duration, TimerMode::Repeating);
+    move |time: Res<Time>| {
+        timer.tick(time.delta());
+        timer.just_finished()
```
Comment (mockersf) [inline crates/bevy_time/src/common_conditions.rs:39]: I think you can do `timer.tick(time.delta()).just_finished()`

**Review by mockersf — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
bors r+

**Pull request closed without merging**